sambacc/grpc/backend.py

```diff
@@ -0,0 +1,161 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>
+#
+
+from typing import Any, Union
+
+import dataclasses
+import json
+import os
+import subprocess
+
+from sambacc.typelets import Self
+import sambacc.config
+import sambacc.samba_cmds
+
+
+@dataclasses.dataclass
+class Versions:
+    samba_version: str = "foo"
+    sambacc_version: str = "bar"
+    container_version: str = "baz"
+
+
+@dataclasses.dataclass
+class Session:
+    session_id: str
+    username: str
+    groupname: str
+    remote_machine: str
+    hostname: str
+    session_dialect: str
+    uid: int
+    gid: int
+
+    @classmethod
+    def load(cls, json_object: dict[str, Any]) -> Self:
+        return cls(
+            session_id=json_object.get("session_id", ""),
+            username=json_object.get("username", ""),
+            groupname=json_object.get("groupname", ""),
+            remote_machine=json_object.get("remote_machine", ""),
+            hostname=json_object.get("hostname", ""),
+            session_dialect=json_object.get("session_dialect", ""),
+            uid=int(json_object.get("uid", -1)),
+            gid=int(json_object.get("gid", -1)),
+        )
+
+
+@dataclasses.dataclass
+class TreeConnection:
+    tcon_id: str
+    session_id: str
+    service_name: str
+
+    @classmethod
+    def load(cls, json_object: dict[str, Any]) -> Self:
+        return cls(
+            tcon_id=json_object.get("tcon_id", ""),
+            session_id=json_object.get("session_id", ""),
+            service_name=json_object.get("service", ""),
+        )
+
+
+@dataclasses.dataclass
+class Status:
+    timestamp: str
+    version: str
+    sessions: list[Session]
+    tcons: list[TreeConnection]
+
+    @classmethod
+    def load(cls, json_object: dict[str, Any]) -> Self:
+        return cls(
+            timestamp=json_object.get("timestamp", ""),
+            version=json_object.get("version", ""),
+            sessions=[
+                Session.load(v)
+                for _, v in json_object.get("sessions", {}).items()
+            ],
+            tcons=[
+                TreeConnection.load(v)
+                for _, v in json_object.get("tcons", {}).items()
+            ],
+        )
+
+    @classmethod
+    def parse(cls, txt: Union[str, bytes]) -> Self:
+        return cls.load(json.loads(txt))
+
+
+class ControlBackend:
+    def __init__(self, config: sambacc.config.InstanceConfig) -> None:
+        self._config = config
+
+    def _samba_version(self) -> str:
+        smbd_ver = sambacc.samba_cmds.smbd["--version"]
+        res = subprocess.run(list(smbd_ver), check=True, capture_output=True)
+        return res.stdout.decode().strip()
+
+    def _sambacc_version(self) -> str:
+        try:
+            import sambacc._version
+
+            return sambacc._version.version
+        except ImportError:
+            return "(unknown)"
+
+    def _container_version(self) -> str:
+        return os.environ.get("SAMBA_CONTAINER_VERSION", "(unknown)")
```

Comment:
What does `SAMBA_CONTAINER_VERSION` contain?

Reply:
Right now nothing. My plan was that once this is available in sambacc to add it as an _optional_ build var to samba container. Then any "vendor" can set it to some helpful information or just leave it blank.

Reply:
In that case I'm not sure if `SAMBA_CONTAINER_VERSION` is the right name. At the same time  I can't think of anything that suits your thoughts. I think we can leave it as it is.

Reply:
Yup. naming is hard. :-)